### ISSUE TITLE
Type mismatch

### ISSUE THREAD

**Issue by mogusbi-motech** [kripod/react-polymorphic-box#25]
## Description

Creating polymorphic components generates TypeScript errors. It appears the TypeScript compiler doesn't like it when you pass in `...restProps` if it contains the `as` prop. If you remove that it no longer complains so it appears there is some kind of mismatch for the `as` type.

## Reproduction

Create the `Heading` component example from the README.

## Expected behavior

No TypeScript errors

## Actual behavior

<img width="1057" alt="image" src="https://github.com/kripod/react-polymorphic-box/assets/53015240/77ec666e-f702-498e-9fc5-9c2573dd6959">

```
Type '{ as: "h2"; style: { color: string | undefined; } & ComponentProps<E>[string]; } & Omit<HeadingProps<E>, "style" | "color">' is not assignable to type 'IntrinsicAttributes & BoxOwnProps<"h2" & HeadingProps<E>["as"]> & Omit<ComponentProps<"h2" & HeadingProps<E>["as"]>, "as">'.
Type '{ as: "h2"; style: { color: string | undefined; } & ComponentProps<E>[string]; } & Omit<HeadingProps<E>, "style" | "color">' is not assignable to type 'Omit<ComponentProps<"h2" & HeadingProps<E>["as"]>, "as">'.
```

## System:
 - OS: macOS 13.4
 - CPU: (16) x64 Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
 - Memory: 1.07 GB / 32.00 GB
 - Shell: 5.9 - /bin/zsh
## Binaries:
 - Node: 18.15.0 - ~/.nvm/versions/node/v18.15.0/bin/node
 - Yarn: 3.5.1 - ~/.nvm/versions/node/v18.15.0/bin/yarn
 - npm: 8.19.4 - ~/Repos/motech-development/platform/node_modules/.bin/npm
## Browsers:
 - Chrome: 113.0.5672.126
 - Edge: 113.0.1774.57
 - Firefox: 109.0.1
 - Safari: 16.5

**Comment by mogusbi-motech** [comment]
Additional info that may or may not be relevant;

React: 18.2.0 (types: 18.2.8)
React DOM: 18.2.0 (types: 18.2.4)
Typescript: 5.0.4

**Comment by kripod** [comment]
Thank you for reporting.

Unfortunately, I wouldn’t recommend using either this project or [react-polymorphic-types](https://github.com/kripod/react-polymorphic-types) anymore.

I’d suggest specifying a `render` prop instead of `as` as follows:

```tsx
import clsx from "clsx";
import * as React from "react";

export type ButtonProps = React.ComponentPropsWithRef<"button"> & {
  render?: (
    props: React.ComponentPropsWithoutRef<"button"> &
      Required<
        Pick<
          React.ComponentPropsWithoutRef<"button">,
          "disabled" | "className" | "children"
        >
      >,
  ) => React.ReactNode;
};

export const Button = React.forwardRef(function Button(
  { disabled, className, children = null, render, ...restProps }: ButtonProps,
  ref: React.ForwardedRef<HTMLButtonElement>,
) {
  return (
    <>
      {(render ?? ((props) => <button ref={ref} type="button" {...props} />))({
        disabled: Boolean(disabled),
        className: clsx(
          /* Base class names may be added here */
          className,
        ),
        children,
        ...restProps,
      })}
    </>
  );
});
```

Usage:

```tsx
<Button
  render={renderButtonAsLink(({ className, children }) => (
    <a href="#_" className={className}>
      {children}
    </a>
  ))}
>
  Click me
</Button>

function renderButtonAsLink(
  render: (
    props: React.ComponentPropsWithoutRef<"button"> &
      Required<
        Pick<React.ComponentPropsWithoutRef<"button">, "className" | "children">
      >,
  ) => React.ReactNode,
) {
  return (({ disabled, className, children, ...restProps }) =>
    disabled ? (
      // https://www.scottohara.me/blog/2021/05/28/disabled-links.html
      // eslint-disable-next-line jsx-a11y/anchor-is-valid
      <a role="link" aria-disabled className={className}>
        {children}
      </a>
    ) : (
      render({ disabled, className, children, ...restProps })
    )) satisfies ButtonProps["render"];
}
```

**Comment by nsaunders** [comment]
@kripod thanks for your past work on react-polymorphic-box and for the update regarding render props. Would you happen to be able to link me to any reference material (such as a blog post) that would explain your particular implementation of render props further?

**Comment by kripod** [comment]
@nsaunders Unfortunately, I don’t really find the time to create a blog, even though I could write quite a bit about the topic. I’m not aware of any articles mentioning render props to solve polymorphism.

**Comment by nsaunders** [comment]
@kripod thanks, I really appreciate the response. If there's any chance you could just write down 1-3 quick bullet points here on your motivation for moving away from the `as` prop, I would appreciate it tremendously. To me, render props seem like the best type-safe alternative to the `as` prop, but I have to admit that I don't fully understand why everyone is "breaking up with the `as` prop", so to speak, in the first place. At the same time, I have high-impact architectural decisions to make regarding v2 of a component library and am trying to eliminate as many potential blind spots for myself as possible. Either way, thanks again: Your efforts here have helped me out a lot over the past couple of years.

**Comment by kripod** [comment]
Apologies for my late response.

The main reason I’d go with a render prop is that the props passed to the rendered element should be mapped in most cases. For instance, while a polymorphic `Button` component may have a `disabled` prop that can be passed to a `<button>` element,  the same prop can’t be passed to an `<a>`. Therefore, we have to map the props to be passable to `<a>` [as shown above](https://github.com/kripod/react-polymorphic-box/issues/25#issuecomment-1575558809).

Also, the reason I haven’t adopted [the `Slot` approach of Radix UI](https://www.radix-ui.com/docs/primitives/utilities/slot) is that sometimes, a polymorphic component may want to modify the `children` it passes for rendering. E.g. instead of just rendering the `children` retrieved from the outside, `<Button disabled="loading">` may render `<span className={/* … */}>{children} <Spinner /></span>` instead of just the original `children`. The same behavior applies to making `ButtonProps["children"]` a render prop: we can’t extend `children` with a spinner when doing that.

**Comment by JGioia** [comment]
Hi @kripod I encountered this issue a few weeks ago, and I wasn't a huge fan of the render prop or the slot approach, so I adapted this package to separate out the props being sent to the `as` component from the props sent to the polymorphic component. I made another package with these changes that I think fix the issues with this package in Typescript 5 while trying to minimize the number of changes. Here's the git repo: https://github.com/JGioia/react-polymorphic-component/tree/main. Open if you have any comments or want to adapt this approach to your own repo!
